### PR TITLE
Fix lint run in framework and lint source/class/qx/test/*

### PR DESCRIPTION
**Pull request by jochumdev**
This is a follow up of #9154.

I couldn't fix these warnings, please some else fix that:
- Warning: qx.test.ui.basic.Atom (50, 25): Constant as parameter name in function declaration
- Warning: qx.test.Annotation (40, 25): Reference values are shared across all instances: '@methodA'
- Warning: qx.test.Annotation (66, 23): Reference values are shared across all instances: '@methodB'
- Warning: qx.test.Annotation (87, 23): Reference values are shared across all instances: '@methodB'
- Warning: qx.test.Annotation (163, 23): Reference values are shared across all instances: '@methodA'
- Warning: qx.test.bom.Attribute (163, 55): qx.core.Environment.select: second parameter is not a map.
- Warning: qx.test.Mixin (243, 15): Using an undeclared private class feature: '__b'
- Warning: qx.test.Mixin (244, 27): Using an undeclared private class feature: '__b'
- Warning: qx.test.Class (398, 11): Using an undeclared private class feature: '__p'
- Warning: qx.test.Class (406, 20): Using an undeclared private class feature: '__p'
- Warning: qx.test.Class (465, 11): Using an undeclared private class feature: '__p'
- Warning: qx.test.Class (473, 20): Using an undeclared private class feature: '__p'

### CODE REVIEW

**Comment by pcdummy** [comment]
Think that Travis error isn't my fault, right?

**Comment by level420** [comment]
Restarting travis run.
Good night.

**Comment by level420** [comment]
@pcdummy the ```Using an undeclared private class feature:``` could be fixed by declaring the member variable within in the ```members``` part. For the first point in your list in qx.test.Mixin this would be adding

```javascript
      qx.Mixin.define("qx.MPatch",
      {
        members :
        {
          // ADD THIS HERE:
          __b : null,

          sayJuhu : function() { return this.base(arguments) + " Kinners";},

          foo : function(dontRecurs)
          {
            var s = "";
            if (!dontRecurs) {
              this.__b = new qx.Patch2();
              s += "++" + this.__b.foo(true) + "__";
            }

            s += this.base(arguments);
            return s;
          }
        }
      });
```

**Comment by pcdummy** [comment]
Same as with the others, easy. Thanks.

**Comment by level420** [comment]
@pcdummy : the very first lint warning in your list is exactly the situation which caused all this. The test was fixed with https://github.com/qooxdoo/qooxdoo/pull/9152

**Comment by pcdummy** [comment]
Will try to rebase this evening.

**Comment by level420** [comment]
@pcdummy : I think ```Warning: qx.test.bom.Attribute (163, 55): qx.core.Environment.select: second parameter is not a map.``` is, because the generator does not know in this situation that ```setUp``` where __maxLengthValues is initialized, is executed before ```testRemoveAttribute```.
So it may help to assign ```this.__maxLengthValues``` to a local variable before using it as a parameter for ```qx.core.Environment.select```.
Or do something like
```javascript
qx.core.Environment.select("engine.name", (this.__maxLengthValues||{}));
```

**Comment by level420** [comment]
@pcdummy forget it what I've wrote in the comment before does not work. Still investigating.

**Comment by level420** [comment]
No I've understood: as there is explicitly declared int the doc header of qx.core.Environment.select that the second parameter has to be a map, we have to write:
```javascript
    testRemoveAttribute : function()
    {
      var Attribute = qx.bom.element.Attribute;

      Attribute.set(this._input, "maxLength", 10);
      Attribute.set(this._input, "maxLength", null);

      var maxLengthValue = qx.core.Environment.select("engine.name", {
        "mshtml"  : this.__maxLengthValues["mshtml"],
        "webkit"  : this.__maxLengthValues["webkit"],
        "default" : this.__maxLengthValues["default"]
      });
...
```
for that line

**Comment by johnspackman** [comment]
Or just disable the warning for that function?  See http://www.qooxdoo.org/5.1/pages/development/api_jsdoc_ref.html#lint but you should be able to do something like:
````
/**
 * @ignore environmentNonLiteralKey()
 */
testRemoveAttribute : function() {
```

**Comment by pcdummy** [comment]
It doesn't work sadly John :/

**Comment by thron7** [comment]
I think `@ignore environmentNonLiteralKey()` really does only apply to the environment **key**, i.e. the first argument like *"engine.name"* in this case. I'm not sure there is an `@ignore` option for the second, map-type argument in *qx.core.Environment.select()*.

**Comment by level420** [comment]
@pcdummy  and we could also close https://github.com/qooxdoo/qooxdoo/pull/9154 ?

**Comment by pcdummy** [comment]
@level420 yes, this pr incorporates that one.

**Comment by level420** [comment]
@pcdummy @oetiker had that idea. Please change the test to:
```javascript
      qx.Mixin.define("qx.MPatch",
      {
        members :
        {
          sayJuhu : function() { return this.base(arguments) + " Kinners";},

          foo : function(dontRecurs)
          {
            var s = "";
            if (!dontRecurs) {
              this['__b'] = new qx.Patch2();
              s += "++" + this['__b'].foo(true) + "__";
            }

            s += this.base(arguments);
            return s;
          }
        }
      });
```
Acessing this['__b'] instead of this.__b is the trick!

**Comment by pcdummy** [comment]
Integrated that one, hopefully Travis likes us now. Thanks a lot @level420

**Comment by level420** [comment]
:ballot_box_with_check: CALL FOR VOTE

**Comment by derrell** [comment]
This is a huge change -- 91 files -- and many of the changes are whitespace changes which make it really difficult to see what's really relevant...

File: framework/source/class/qx/test/Interface.js [hunk 1]
```diff
@@ -365,7 +363,7 @@ qx.Class.define("qx.test.Interface",
             members :
             {
               getValue : function() {},
-              setValue : function(value) {}
```
Comment (derrell) [inline framework/source/class/qx/test/Interface.js]: Really? A setValue function that doesn't take a value? If lint is complaining about value being unused, this is the wrong fix. If the function really needn't do anything, then I'd declare it as:
```
setvalue : function(value) { value = value; }
```
(Then we'll see if lint complains about a "useless statement"...)

File: framework/source/class/qx/test/Mixin.js [hunk 2]
```diff
@@ -240,8 +240,8 @@ qx.Class.define("qx.test.Mixin",
           {
             var s = "";
             if (!dontRecurs) {
-              this.__b = new qx.Patch2();
-              s += "++" + this.__b.foo(true) + "__";
+              this['__b'] = new qx.Patch2();
```
Comment (derrell) [inline framework/source/class/qx/test/Mixin.js]: This looks like it's trying to fix a "__b is not a member" problem because __b is not declared in the members section?

File: framework/source/class/qx/test/bom/webfonts/Manager.js [hunk 3]
```diff
@@ -89,6 +69,30 @@ qx.Class.define("qx.test.bom.webfonts.Manager", {
       this.__nodesBefore = document.body.childNodes.length;
       this.__sheetsBefore = document.styleSheets.length;
       this.__manager = qx.bom.webfonts.Manager.getInstance();
+
+      this.__fontDefinitions = {
+        finelinerScript :
+        {
```
Comment (derrell) [inline framework/source/class/qx/test/bom/webfonts/Manager.js]: Is __findRule called only once? If __fontDefinitions needs to be different for different instances, then typically the initialization would be done in the constructor.

File: framework/source/class/qx/test/Mixin.js (hunk 2 above)
Comment (level420) [inline framework/source/class/qx/test/Mixin.js]: If you declare __b as member variable then the patch mechanism ```qx.Class.include(qx.Patch1, qx.MPatch)``` complains about __b is not allowed to be overridden.

File: framework/source/class/qx/test/core/Validation.js [hunk 4]
```diff
@@ -145,10 +160,18 @@ qx.Class.define("qx.test.core.Validation",
       var model = this.__model;
 
       // test some wrong inputs
-      this.assertException(function() {model.setString(1);}, qx.core.ValidationError, null, "A number is not a string!");
-      this.assertException(function() {model.setString(this);}, qx.core.ValidationError, null, "This test is not a string!");
-      this.assertException(function() {model.setString(true);}, qx.core.ValidationError, null, "A boolean is not a string!");
-      this.assertException(function() {model.setString(new Date());}, qx.core.ValidationError, null, "A Date-Object is not a string!");
+      this.assertException(
```
Comment (derrell) [inline framework/source/class/qx/test/core/Validation.js]: Yes! Get rid of those excessively long lines!!! Thank you!

File: framework/source/class/qx/test/core/Validation.js [hunk 5]
```diff
@@ -221,9 +259,12 @@ qx.Class.define("qx.test.core.Validation",
       var model = this.__model;
 
       // test some wrong inputs (allowed are male and female)
-      this.assertException(function() {model.setArray(0.999999999);}, qx.core.ValidationError, null, "A 0.999999999 is not in ['male', 'female']");
-      this.assertException(function() {model.setArray("malle");}, qx.core.ValidationError, null, "'malle' is not in ['male', 'female']");
-      this.assertException(function() {model.setArray("");}, qx.core.ValidationError, null, "A empty string is not in ['male', 'female']");
+      this.assertException(
+        function() {model.setArray(0.999999999);}, qx.core.ValidationError, null, "A 0.999999999 is not in ['male', 'female']");
```
Comment (derrell) [inline framework/source/class/qx/test/core/Validation.js]: Sigh... But here those excessively long lines remain. :-(

File: framework/source/class/qx/test/io/part/MockPackage.js [hunk 6]
```diff
@@ -4,13 +4,16 @@ qx.Bootstrap.define("qx.test.io.part.MockPackage",
   {
     this.id = id;
     this.delay = delay || 0;
-    this.error = !!error;
+    this.error = Boolean(error);
```
Comment (derrell) [inline framework/source/class/qx/test/io/part/MockPackage.js]: Why this change?

Comment (pcdummy) [inline framework/source/class/qx/test/io/part/MockPackage.js]: eslint. :)

File: framework/source/class/qx/test/ui/form/AbstractVirtualBox.js [hunk 7]
```diff
@@ -21,12 +21,15 @@ qx.Class.define("qx.test.ui.form.AbstractVirtualBox",
...
+    __selectBox : null,
+
     setUp : function()
     {
-      this.__selectBox = new qx.ui.form.VirtualSelectBox;
+      this.__selectBox = new qx.ui.form.VirtualSelectBox();
       this.getRoot().add(this.__selectBox);
 
-      this.__comboBox = new qx.ui.form.VirtualComboBox;
+      this.__comboBox = new qx.ui.form.VirtualComboBox();
```
Comment (derrell) [inline framework/source/class/qx/test/ui/form/AbstractVirtualBox.js]: Yes. Much more readable.

File: framework/source/class/qx/test/Interface.js (hunk 1 above)
Comment (level420) [inline framework/source/class/qx/test/Interface.js]: @pcdummy does the generator linter really complain here? A few lines below there is still an empty function with parameter value : https://github.com/pcdummy/qooxdoo/blob/297d25359f857c4d4396fbbe5a69e7f60d4bbd36/framework/source/class/qx/test/Interface.js#L401

File: framework/source/class/qx/test/ui/tree/virtual/WidgetProvider.js [hunk 8]
```diff
@@ -286,7 +286,7 @@ qx.Class.define("qx.test.ui.tree.virtual.WidgetProvider",
       var newWidget2Bindungs = widget2.getBindings().length;
       var newModelBindungs = this.getLookupTable().getBindings().length;
       this.assertEquals(oldWidget1Bindungs, newWidget1Bindungs, "Binding on first widget is not removed!");
-      this.assertEquals(oldWidget1Bindungs, newWidget1Bindungs, "Binding on second widget is not removed!");
+      this.assertEquals(oldWidget2Bindungs, newWidget2Bindungs, "Binding on second widget is not removed!");
```
Comment (derrell) [inline framework/source/class/qx/test/ui/tree/virtual/WidgetProvider.js]: Shouldn't we, while we're fixing these lines, fix the spelling of Bindings in this function???

File: framework/source/class/qx/test/util/DateFormat.js [hunk 9]
```diff
@@ -22,35 +22,37 @@ qx.Class.define("qx.test.util.DateFormat",
...
-    {'date' : new Date(-20,10,14), 'result' : {'era': {'abbrev': 'BC', 'fullName': 'Before Christ', 'narrow': 'B'}}},
-    {'date' : new Date(2012, 4, 24, 11, 49, 57, 1), 'result' : {}},
-    {'date' : new Date(2012, 4, 24, 11, 49, 57, 12), 'result' : {}},
-    {'date' : new Date(2012, 4, 24, 11, 49, 57, 123), 'result' : {}}
-
-  ],
+    __dates : null,
+
+    setUp : function() {
+      // result contain an object with what should be expected, the result to test the date against
```
Comment (derrell) [inline framework/source/class/qx/test/util/DateFormat.js]: Once again... Confirm this function is called only once. Typically such initialization is done in the constructor.

**Comment by derrell** [comment]
Ok. I've gone through the whole thing and added a number of line comments. I don't think I'm quite ready to vote on this yet. My comments should probably be addressed first.

**Comment by derrell** [comment]
BTW, this was a huge job. Thanks for doing it!

File: framework/source/class/qx/test/util/DateFormat.js (hunk 9 above)
Comment (level420) [inline framework/source/class/qx/test/util/DateFormat.js]: @derrell but do we have a constructor in the test classes?

**Comment by level420** [comment]
@derrell added some comments to your line notes ;-)

File: framework/source/class/qx/test/util/DateFormat.js (hunk 9 above)
Comment (derrell) [inline framework/source/class/qx/test/util/DateFormat.js]: Maybe not, and maybe this is the right place for it. I just wanted confirmation that this function is called only once, and that the data given it at different times varies. If it's invariant, then the original code -- granted, a reference type initialized in the members section -- is a fine place to put it to avoid multiple assignments. If this setup function is called only once, this is also a fine place to put it, particularly if there's no constructor available in which to put it.

File: framework/source/class/qx/test/Mixin.js (hunk 2 above)
Comment (derrell) [inline framework/source/class/qx/test/Mixin.js]: That's a problem. __b is a private variable. Private variables are not accessible even to mixins, IIRC. The __b in MPatch is different than the __b in Patch1.

File: framework/source/class/qx/test/Interface.js (hunk 1 above)
Comment (pcdummy) [inline framework/source/class/qx/test/Interface.js]: No ESLint complained (i used ESLint to fix those files and run the generator after).

File: framework/source/class/qx/test/util/DateFormat.js (hunk 9 above)
Comment (pcdummy) [inline framework/source/class/qx/test/util/DateFormat.js]: I see setUp as constructor for tests, yes.

File: framework/source/class/qx/test/bom/webfonts/Manager.js (hunk 3 above)
Comment (pcdummy) [inline framework/source/class/qx/test/bom/webfonts/Manager.js]: This is part of "setUp" not "__findRule", the diff isn't clear enough.

File: framework/source/class/qx/test/core/Validation.js (hunk 5 above)
Comment (pcdummy) [inline framework/source/class/qx/test/core/Validation.js]: We should start another Lint with formatting after.

File: framework/source/class/qx/test/ui/tree/virtual/WidgetProvider.js (hunk 8 above)
Comment (pcdummy) [inline framework/source/class/qx/test/ui/tree/virtual/WidgetProvider.js]: New patch with that will come.

File: framework/source/class/qx/test/core/Validation.js (hunk 5 above)
Comment (thron7) [inline framework/source/class/qx/test/core/Validation.js]: From experience I can only recommend to put formatting changes in "whitespace-only" commits. When you mix them with real changes it is very hard to review those commits.

**Comment by level420** [comment]
@derrell please ad a dissenting vote to this PR to avoid someone is accidentally merging it.

**Comment by derrell** [comment]
Down-voted to avoid PR being merged prematurely.

File: framework/source/class/qx/test/util/DateFormat.js (hunk 9 above)
Comment (level420) [inline framework/source/class/qx/test/util/DateFormat.js]: @derrell I think this change is OK as (as stated in the docs) creating the array with test dates again for each test does not harm and each test has reliable the same initial values.

File: framework/source/class/qx/test/bom/webfonts/Manager.js (hunk 3 above)
Comment (level420) [inline framework/source/class/qx/test/bom/webfonts/Manager.js]: Yes. Seems reasonable to have a fresh initialized testdata array for each test.

File: framework/source/class/qx/test/Mixin.js (hunk 2 above)
Comment (level420) [inline framework/source/class/qx/test/Mixin.js]: @derrell The fact is if we declare it as a member, then the exception https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/test/Mixin.js#L258 is thrown. But why? The class qx.Patch1 where qx.MPatch is included, does not have a private member __b itself. And, as discussed with @oetiker private members are valid in mixins.

What if in the case of the test, the intention was not to write clean code, but trick the testsuite somehow to be able to implement that test? In this case it may be valid to have a private variable not be declared member and hide it from the linter via a @lint hint.

We could merger @thron7 's PR https://github.com/qooxdoo/qooxdoo/pull/9164 first, leave the code as it was, asking @pcdummy t revert this changes and to rebase this PR after merging https://github.com/qooxdoo/qooxdoo/pull/9164, which also includes the fix by adding the lint hint.

File: framework/source/class/qx/test/Interface.js (hunk 1 above)
Comment (level420) [inline framework/source/class/qx/test/Interface.js]: @pcdummy I think @derrell is right here. As the test is to check if a class fulfills the requirements of the interface it would be valid to have some code in the implementing method. Even such code like
```javascript
value += '';
```

**Comment by level420** [comment]
@derrell please re-evaluate this PR. Thank you.

File: framework/source/class/qx/test/util/DateFormat.js (hunk 9 above)
Comment (derrell) [inline framework/source/class/qx/test/util/DateFormat.js]: Agree. The change is fine.

**Comment by derrell** [comment]
I think all but one of my questions have been answered. The one pending is setvalue(value). Excluding optional parameters is common practice, but mandatory parameters should always be listed even if unused. If that means we have to "use" them with some trivial, idempotent "change" to shut up the linter, then that's what we should do.

**Comment by level420** [comment]
@pcdummy would you please be so kind and change the PR discussed in https://github.com/qooxdoo/qooxdoo/pull/9158#discussion_r77768439 and above? 
Thank you.

**Comment by level420** [comment]
@pcdummy merget https://github.com/qooxdoo/qooxdoo/pull/9164

**Comment by level420** [comment]
@pcdummy do you have some time to work again on this PR?

After some back and forth and discussion with @derrell it would be good if you can change the private variable __b to a protected one _b and add it to the members of qx.MPatch ( in https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/test/Mixin.js#L233-L280 )
then I'll start a new voting round.

That would be great!

**Comment by pcdummy** [comment]
@level420 all done if travis is will be green.

**Comment by level420** [comment]
@pcdummy Thank you.

But I can't find the change ```qx.test.Mixin.testPatchOverwritten``` where the private member ```__b``` is changed to a protected one ```_b``` and is explicitly declared as a member ( https://github.com/pcdummy/qooxdoo/blob/fix-lint-constants/framework/source/class/qx/test/Mixin.js#L244 ).

There is also missing the change where in https://github.com/pcdummy/qooxdoo/blob/fix-lint-constants/framework/source/class/qx/test/Interface.js#L380 the ```setValue``` method should have a parameter again, maybe with some dummy code in the function body like ``` value = value; ``` to make eslint happy.

At https://github.com/pcdummy/qooxdoo/blob/fix-lint-constants/framework/source/class/qx/test/Interface.js#L403 there is also such an empty ```setValue(value)``` function declaration. Doesn't eslint complain here as well?

And please note: until we resolve issue https://github.com/qooxdoo/qooxdoo/issues/9167 the travis run will never complete successfully.

**Comment by level420** [comment]
@pcdummy would be nice if you have some time to add the last changes so we can merge and close this PR. Thank you.

**Comment by pcdummy** [comment]
This should be finaly ready to Vote and Merge.

File: framework/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# pcdummy's qooxdoo
```
Comment (cajus) [inline framework/README.md]: So this file may need to be removed, I guess.

File: framework/package.json [hunk 11]
```diff
@@ -1,11 +1,27 @@
 {
```
Comment (cajus) [inline framework/package.json]: This one shouldn't get changes, too for this commit, or am I wrong? ;-)

**Comment by pcdummy** [comment]
Made a badass mistake, rebased on the wrong master branch (mine instead of upstream). Will fix.

File: framework/source/class/qx/test/dev/unit/Sinon.js [hunk 12]
```diff
@@ -235,9 +235,6 @@ qx.Class.define("qx.test.dev.unit.Sinon",
       var func = function() {};
       var obj = {"a": function() {}};
 
-      var spy = this.spy(func);
-      var stub = this.stub(obj, "a");
-      var xhr = this.useFakeXMLHttpRequest();
```
Comment (level420) [inline framework/source/class/qx/test/dev/unit/Sinon.js]: What if this lines test Sinon itself? Of course the variable ```spy``` is not used anymore, but the call ```this.spy(func)``` itself could be the test. If ```this.spy(func)``` fails because of an exception, then the Sinon integration may be faulty.

File: framework/source/class/qx/test/Interface.js [hunk 13]
```diff
@@ -71,8 +71,6 @@ qx.Class.define("qx.test.Interface",
         properties : { color : { } }
       });
 
-      var audi = new qx.test.i.Audi("audi");
```
Comment (level420) [inline framework/source/class/qx/test/Interface.js]: I don't know if we should delete this line. The instantiation itself could fail with an exception which then would maybe be a test itself.

File: framework/source/class/qx/test/Interface.js [hunk 14]
```diff
@@ -173,7 +171,7 @@ qx.Class.define("qx.test.Interface",
       if (this.isDebugOn())
       {
         this.assertException(function() {
-          var i = new qx.test.i.ICar();
+          qx.test.i.ICar();
```
Comment (level420) [inline framework/source/class/qx/test/Interface.js]: Is this really equivalent? Does simply writing ```qx.text.i.ICar()``` create a new instance of that class? I'm asking because I really don't know that.

File: framework/source/class/qx/test/io/rest/Resource.js [hunk 15]
```diff
@@ -228,7 +223,7 @@ qx.Class.define("qx.test.io.rest.Resource",
       this.require(["debug"]);
 
       this.assertException(function() {
-        var res = new qx.io.rest.Resource([]);
+        qx.io.rest.Resource([]);
       });
```
Comment (level420) [inline framework/source/class/qx/test/io/rest/Resource.js]: Here again: is ```qx.io.rest.Resource([]);``` really equivalent to ```new qx.io.rest.Resource([]);```?
I don't thinks so.

File: framework/source/class/qx/test/mobile/core/Widget.js [hunk 16]
```diff
@@ -136,7 +136,7 @@ qx.Class.define("qx.test.mobile.core.Widget",
       if (qx.core.Environment.get("qx.debug"))
       {
         this.assertException(function() {
-          var widget2 = new qx.ui.mobile.core.Widget().set({id:"affe"});
+          qx.ui.mobile.core.Widget().set({id:"affe"});
         });
```
Comment (level420) [inline framework/source/class/qx/test/mobile/core/Widget.js]: And again: Is calling ```qx.ui.mobile.core.Widget().set({id:"affe"});``` equivalent to ```new qx.ui.mobile.core.Widget().set({id:"affe"});```?

File: framework/source/class/qx/test/theme/simple/Appearance.js [hunk 17]
```diff
@@ -443,13 +443,6 @@ qx.Class.define("qx.test.theme.simple.Appearance",
       this.assertIdentical(qx.theme.simple.Image.URLS["treevirtual-cross"], style.icon);
     },
 
-    testTreeVirtualCross : function()
```
Comment (level420) [inline framework/source/class/qx/test/theme/simple/Appearance.js]: Why is this test completely removed?

**Review by level420 — CHANGES_REQUESTED** [review]

**Comment by level420** [comment]
@pcdummy it would be nice if you have some time to finish this work. it would be a great pity if all this work will we lost! thank you.

**Comment by cboulanger** [comment]
@pcdummy Another year has passed and we still haven't merged your PR, after all these iterations. Would you be so kind to update your branch?

**Comment by cboulanger** [comment]
No feedback for over a month, closing,
